distributed/client.py

```diff
@@ -807,6 +820,135 @@ class VersionsDict(TypedDict):
     client: dict[str, dict[str, Any]]
 
 
+_T_LowLevelGraph: TypeAlias = dict[Key, tuple]
+
+
+def _is_nested(iterable):
+    for item in iterable:
+        if (
+            isinstance(item, Iterable)
+            and not isinstance(item, str)
+            and not isinstance(item, bytes)
+        ):
+            return True
+    return False
+
+
+class _MapLayer(Layer):
+    func: Callable
+    iterables: Iterable[Any]
+    key: str | Iterable[str] | None
+    pure: bool
+    annotations: dict[str, Any] | None
+
+    def __init__(
+        self,
+        func: Callable,
+        iterables: Iterable[Any],
+        key: str | Iterable[str] | None = None,
+        pure: bool = True,
+        annotations: dict[str, Any] | None = None,
+        **kwargs,
+    ):
+        self.func: Callable = func
+        self.iterables: Iterable[Any] = (
+            list(zip(*zip(*iterables))) if _is_nested(iterables) else [iterables]
+        )
+        self.key: str | Iterable[str] | None = key
+        self.pure: bool = pure
+        self.kwargs = kwargs
+        super().__init__(annotations=annotations)
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__} <func='{funcname(self.func)}'>"
+
+    @property
+    def _dict(self) -> _T_LowLevelGraph:
+        self._cached_dict: _T_LowLevelGraph
+        dsk: _T_LowLevelGraph
+
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    @property
```

Comment:
This should be cached. The keys are generating UUIDs (pure=false) but this has to return the same value every time